### PR TITLE
IDEMPIERE-5567 Support of UUID as Key (FHCA-4195) - ChangeLog - FKRules

### PR DESCRIPTION
**Pull request by CarlosRuiz-globalqss**
https://idempiere.atlassian.net/browse/IDEMPIERE-5567

- implement ChangeLog for UUID Based Tables
- implement foreign key rules for Record_UU fields

# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [x] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] I have checked my code and corrected any misspellings
- [x] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
### Tests
- [x] I have tested the direct scenario that my code is solving
- [ ] I checked all collaterals that can be affected by my changes, and tested other potential affected scenarios
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added unit tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: org.adempiere.base/src/org/compiere/model/PO_Record.java [hunk 1]
```diff
@@ -59,46 +60,63 @@
...
 		//	Parent Loop
 		for (int i = 0; i < s_parents.length; i++)
 		{
 			if (s_parents[i] == AD_Table_ID)
 			{
 				int AD_Table_IDchild = s_parentChilds[i];
+				StringBuilder whereClauseChild = new StringBuilder(" AD_Table_ID=? AND ").append(columnName).append(" IN (SELECT ")
+						.append(s_parentChildNames[i]).append("_ID FROM ")
+						.append(s_parentChildNames[i]).append(" WHERE ")
+						.append(s_parentNames[i]).append("_ID=?)");
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO_Record.java:115]: this is inconsistent. if the change is to support UUID only table, you can't assume parent is using _ID.

File: org.adempiere.base/src/org/compiere/model/PO_Record.java [hunk 2]
```diff
@@ -59,46 +60,63 @@
...
+		String columnName;
+		if (Record_IDorUU instanceof Integer) {
+			refId = DisplayType.RecordID;
+			columnName = "Record_ID";
+		} else if (Record_IDorUU instanceof String) {
+			refId = DisplayType.RecordUU;
+			columnName = "Record_UU";
+		} else {
+			log.warning(Record_IDorUU.getClass().getName() + " not supported for ID/UUID");
+			return true;
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO_Record.java:79]: better throw illegalargumentexception here. should never happens.

File: org.adempiere.base/src/org/compiere/model/PO_Record.java [hunk 3]
```diff
@@ -109,43 +127,68 @@ static boolean deleteRecordIdCascade (int AD_Table_ID, int Record_ID, String trx
...
-		KeyNamePair[] tables = getTablesWithModelCascade(tableName, trxName);
+	public static void deleteModelCascade(String tableName, Serializable Record_IDorUU, String trxName) {
+		int refId;
+		if (Record_IDorUU instanceof Integer) {
+			refId = DisplayType.RecordID;
+		} else if (Record_IDorUU instanceof String) {
+			refId = DisplayType.RecordUU;
+		} else {
+			log.warning(Record_IDorUU.getClass().getName() + " not supported for ID/UUID");
+			return;
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO_Record.java:152]: better throw illegalargumentexception here. should never happens.

File: org.adempiere.base/src/org/compiere/model/PO_Record.java [hunk 4]
```diff
@@ -109,43 +127,68 @@ static boolean deleteRecordIdCascade (int AD_Table_ID, int Record_ID, String trx
...
+			refTableDirId = DisplayType.TableDir;
+			refTableId = DisplayType.Table;
+			refTableSearchId = DisplayType.Search;
+		} else if (refId == DisplayType.RecordUU) {
+			refTableDirId = DisplayType.TableDirUU;
+			refTableId = DisplayType.TableUU;
+			refTableSearchId = DisplayType.SearchUU;
+		} else {
+			log.warning(refId + " not supported for ID/UUID");
+			return null;
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO_Record.java:189]: better throw illegalargumentexception here. should never happens.

File: org.adempiere.base/src/org/compiere/model/PO_Record.java [hunk 5]
```diff
@@ -184,80 +228,124 @@ private static KeyNamePair[] getTablesWithModelCascade(String tableName, String
...
+		String columnName;
+		if (Record_IDorUU instanceof Integer) {
+			refId = DisplayType.RecordID;
+			columnName = "Record_ID";
+		} else if (Record_IDorUU instanceof String) {
+			refId = DisplayType.RecordUU;
+			columnName = "Record_UU";
+		} else {
+			log.warning(Record_IDorUU.getClass().getName() + " not supported for ID/UUID");
+			return;
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO_Record.java:245]: better throw illegalargumentexception here. should never happens.

File: org.adempiere.base/src/org/compiere/model/PO_Record.java [hunk 6]
```diff
@@ -184,80 +228,124 @@ private static KeyNamePair[] getTablesWithModelCascade(String tableName, String
...
 	/**
 	 * 	An entry Exists for restrict table/record combination
 	 *	@param AD_Table_ID table
-	 *	@param Record_ID record
+	 *	@param Record_IDorUU record ID (int) or UUID (String)
 	 *	@param trxName transaction
 	 *	@return error message (Table Name) or null
 	 */
-	static String exists (int AD_Table_ID, int Record_ID, String trxName)
+	static String exists (int AD_Table_ID, Serializable Record_IDorUU, String trxName)
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO_Record.java:287]: this is using package visibility. better change it to either protected or public

**Comment by hengsin** [comment]
hi @CarlosRuiz-globalqss , considering the size of UUID and the volume of change log, the storage implication of this change  could be huge.

File: org.adempiere.base/src/org/compiere/model/PO_Record.java (hunk 2 above)
Comment (CarlosRuiz-globalqss) [reply org.adempiere.base/src/org/compiere/model/PO_Record.java]: done

File: org.adempiere.base/src/org/compiere/model/PO_Record.java (hunk 4 above)
Comment (CarlosRuiz-globalqss) [reply org.adempiere.base/src/org/compiere/model/PO_Record.java]: done

File: org.adempiere.base/src/org/compiere/model/PO_Record.java (hunk 6 above)
Comment (CarlosRuiz-globalqss) [reply org.adempiere.base/src/org/compiere/model/PO_Record.java]: done

File: org.adempiere.base/src/org/compiere/model/PO_Record.java (hunk 1 above)
Comment (CarlosRuiz-globalqss) [reply org.adempiere.base/src/org/compiere/model/PO_Record.java:114]: I'm debugging this - it is a strange code anyways.

The deletion of orderlines used to happen on MOrder.beforeDelete - and I moved that to Model Cascade with ticket IDEMPIERE-2060

There is a bug as you noticed, but the code looks weird, I think is unnecessary.

Comment (hengsin) [reply org.adempiere.base/src/org/compiere/model/PO_Record.java:114]: I've not inspect PO_Record before this, very strange one indeed.

**Comment by CarlosRuiz-globalqss** [comment]
> hi @CarlosRuiz-globalqss , considering the size of UUID and the volume of change log, the storage implication of this change could be huge.

Ready @hengsin - added the SysConfig variable AD_CHANGELOG_SAVE_UUID, possible values:
- B -> Just for UUID based tables (the default)
- A -> Always
- U -> just UUID not ID

**Review by hengsin — APPROVED** [review]
looks good

**Pull request merged**